packages/utils/src/get-snark-artifacts/get-snark-artifacts.node.ts

```diff
@@ -0,0 +1,98 @@
+import { createWriteStream, existsSync } from "node:fs"
+import { mkdir } from "node:fs/promises"
+import { dirname } from "node:path"
+import os from "node:os"
+import { SnarkArtifacts, ProofType } from "../types"
+import { ARTIFACT_BASE_URLS, ARTIFACTS_TYPES } from "./config"
+
+const getArtifactBaseUrl = (proofType: ProofType, numberOfInputs?: number): string => {
+    if (proofType === ProofType.POSEIDON) {
+        if (numberOfInputs === undefined) {
+            throw new Error("numberOfInputs is required for Poseidon proof")
+        }
+        if (numberOfInputs < 1) {
+            throw new Error("numberOfInputs must be greater than 0")
+        }
+        return `${ARTIFACT_BASE_URLS[proofType]}/${numberOfInputs}`
+    }
+    return ARTIFACT_BASE_URLS[proofType]
+}
+
+async function download(url: string, outputPath: string) {
+    const response = await fetch(url)
+    if (!response.ok) {
+        throw new Error(`Failed to fetch ${url}: ${response.statusText}`)
+    }
+
+    const dir = dirname(outputPath)
+    try {
+        await mkdir(dir, { recursive: true })
+    } catch (error) {
+        throw new Error(`Failed to create directory ${dir}: ${(error as Error).message}`)
+    }
+
+    const fileStream = createWriteStream(outputPath)
+    if (!fileStream) {
+        throw new Error(`Failed to create write stream for ${outputPath}`)
+    }
+
+    const reader = response.body?.getReader()
+    if (!reader) {
+        throw new Error("Failed to get response body reader")
+    }
+
+    try {
+        const pump = async () => {
+            const { done, value } = await reader.read()
+            if (done) {
+                fileStream.end()
+                return
+            }
+
+            fileStream.write(Buffer.from(value))
+            await pump()
+        }
+
+        await pump()
+    } catch (error) {
+        fileStream.close()
+        throw error
+    }
```

Comment:
even if node > 18 has built in fetch. There are still some mismatch between the stream api of node and [whatwg](https://streams.spec.whatwg.org/): node fetch api returns a [ReadableStream](https://streams.spec.whatwg.org/#rs-class) which isn't directly compatible with node stream system. This is why `as any` type casting was necessary.
We can avoid type casting by using the stream reader from the ReadableStream which is part of the WHATWG api.  
It also makes unit testing easier.  
But it is lot more verbose....